### PR TITLE
Parse bounds declarations for function return values.

### PR DESCRIPTION
**Pull request by dtarditi**
This extends parsing of function declarators to parse bounds expressions for
function return values.  A return bounds expression is declared by following
a parameter list with ':' bounds-expression.

The return bounds information is propagated to the representation of function
declarations in clang.  It is not propagated yet to function types because
there is no way to represent it in function types.  I have opened issue #20
to track that and added a few TODO's to the code flagging where the
information will need to be added.

Testing:
- Add feature tests for parsing return bounds.  The tests are in 
the file return_bounds.c, which will be committed separately to the checkedc
Github repo.
- Passes existing clang regression test baseline.

### CODE REVIEW

File: include/clang/Sema/DeclSpec.h [hunk 1]
```diff
@@ -1287,11 +1287,18 @@ struct DeclaratorChunk {
     /// \brief The end location of the exception specification, if any.
     unsigned ExceptionSpecLocEnd;
 
+    /// The location of the ':' for the return bounds expression in the source
+    unsigned ReturnBoundsColon;
```
Comment (reubeno) [inline include/clang/Sema/DeclSpec.h]: >ReturnBoundsColon [](start = 13, length = 17)

[Nit] I'd suggest including 'Loc' in the name to make it clear that this is a location.

File: include/clang/Sema/DeclSpec.h [hunk 2]
```diff
@@ -1352,6 +1359,10 @@ struct DeclaratorChunk {
       return SourceLocation::getFromRawEncoding(RParenLoc);
     }
 
+    SourceLocation getReturnBoundsColon() const {
```
Comment (reubeno) [inline include/clang/Sema/DeclSpec.h]: >getReturnBoundsColon [](start = 19, length = 20)

[Nit] Ditto from above -- suggest including Loc in the name.

File: lib/Parse/ParseDecl.cpp [hunk 3]
```diff
@@ -5741,6 +5747,17 @@ void Parser::ParseFunctionDeclarator(Declarator &D,
     }
   }
 
+  if (getLangOpts().CheckedC && Tok.is(tok::colon)) {
```
Comment (reubeno) [inline lib/Parse/ParseDecl.cpp]: >if (getLangOpts().CheckedC && Tok.is(tok::colon)) { [](start = 2, length = 51)

I know you called it out in the header comment for this function, but a quick brief comment calling out that this is for the return bounds expression might be nice (similar to the ones in the few code chunks directly above).

**Comment by reubeno** [comment]
I called out a few very low-priority suggestions; otherwise, looks great.

**Comment by awruef** [comment]
Looks good to me.

**Pull request merged**